users/dierkes/preprocessing/wav2vec.py

```diff
@@ -60,8 +79,155 @@ def run(self):
+        self.rqmt = {
+            "time": 2,
+            "mem": 32,
+            "cpu": 1,
+            "gpu": 1 if self.device == "gpu" else 0,
+        }
+
+    def tasks(self):
+        yield Task("run", rqmt=self.rqmt)
+
+    def run(self):
+        better_exchook.install()
+        debug_register_better_repr()
+        log.initialize(verbosity=[5])
+        log.Verbosity = 10
+        setup_tf_thread_pools()
+        print_available_devices()
+
+        if self.fairseq_root:
+            sys.path.append(self.fairseq_root.get_path())
+        if self.pytorch_to_returnn_root:
+            sys.path.append(self.pytorch_to_returnn_root.get_path())
+
+        def model_func(wrapped_import, inputs: torch.Tensor):
+            if typing.TYPE_CHECKING or not wrapped_import:
+                import torch
+                import fairseq
+                from fairseq.models import wav2vec as wav2vec_models
+            else:
+                torch = wrapped_import("torch")
+                torch_nn = wrapped_import("torch.nn.modules")
+                fairseq = wrapped_import("fairseq")
+                wav2vec_models = wrapped_import("fairseq.models.wav2vec")
+                # the following imports are necessary to make sure that these parts are registered also for the wrapped runs
+                audio_pretraining = wrapped_import("fairseq.tasks.audio_pretraining")
+                adam = wrapped_import("fairseq.optim.adam")
+                polynomial_decay = wrapped_import(
+                    "fairseq.optim.lr_scheduler.polynomial_decay_schedule"
+                )
+                wav2vec_criterion = wrapped_import(
+                    "fairseq.criterions.wav2vec_criterion"
+                )
+
+            # Initialize PyTorch example
+            model, cfg, task = fairseq.checkpoint_utils.load_model_ensemble_and_task(
+                [self.wav2vec_config.get_path()]
+            )
+            w2v2_model = model[0]
+            print(
+                "WARNING: Simplifications: feature_grad_mult=1.0, encoder layer dropout=0.0"
+            )
+            w2v2_model.feature_grad_mult = 1.0
+            w2v2_model.encoder.layerdrop = 0.0
+            print(
+                "WARNING: remove quantizer, this involves using non-optimized parameters"
+            )
+            w2v2_model.quantizer = None
+            w2v2_model.project_q.in_features = w2v2_model.embed
+            w2v2_model.project_q._parameters["weight"] = torch.from_numpy(
+                np.random.randn(256, 512).astype("float32")
+            )
+
+            print("WARNING: simplify model by setting mask=False")
+            return w2v2_model(inputs, mask=False, features_only=True)["x"]
+
+        print("Load audio")
+        print("  ", self.audio_input.get_path())
+        audio, fr = soundfile.read(self.audio_input.get_path())
+        audio = np.reshape(audio, [4, len(audio) // 4]).astype(
+            "float32"
+        )  # dim (B, T) -> (4, *)
```

Comment:
We could also make this more generic. The input (not necessarily audio btw) could be an input to the job as a numpy array. Then, we'd just do the `torch.from_numpy` here.